src/consumer/stream_consumer.rs

```diff
@@ -49,76 +49,89 @@ unsafe fn disable_nonempty_callback(queue: &NativeQueue) {
     rdsys::rd_kafka_queue_cb_event_enable(queue.ptr(), None, ptr::null_mut())
 }
 
-struct WakerSlab {
-    wakers: Mutex<Slab<Option<Waker>>>,
+/// Data structure to store waker instances from idle streams.
+/// This queue wakes up the wakers in FIFO order.
+/// On registration streams receive an id they can reuse for setting and removing wakers.
+///
+/// Implementation is optimized for adding new wakers and waking all of them.
+/// Removing a waker is less efficient.
+struct WakerQueue {
+    /// Vec containing the set waker instances and the ids of their owning streams
+    waker_queue: Mutex<Vec<(usize, Waker)>>,
+    /// Counts the number of registered clients. This used to generate ids on registration
+    num_registered: AtomicUsize,
 }
 
-impl WakerSlab {
-    fn new() -> WakerSlab {
-        WakerSlab {
-            wakers: Mutex::new(Slab::new()),
+impl WakerQueue {
+    fn new() -> WakerQueue {
+        WakerQueue {
+            waker_queue: Mutex::new(Vec::new()),
+            num_registered: AtomicUsize::new(0),
         }
     }
 
+    /// Awakes all set wakers in order of their arrival
     fn wake_all(&self) {
-        let mut wakers = self.wakers.lock().unwrap();
-        for (_, waker) in wakers.iter_mut() {
-            if let Some(waker) = waker.take() {
-                waker.wake();
-            }
+        let waker_queue = std::mem::take(&mut *self.waker_queue.lock().unwrap());
+        for (_, waker) in waker_queue.into_iter() {
+            waker.wake();
         }
     }
 
+    /// Returns a unique id that can be reused by the caller to set and unset its waker
     fn register(&self) -> usize {
```

Comment:
maybe instead of having additional atomic during register we could simply put placeholder value into the vec instead and return index of such slot? this would allow to avoid having unset_waker being `O(self.waker_queue.len())`.

Reply:
I like the idea in general. But it makes everything a bit too complicated in my opinion. We'd need to ensure that we always unset the stored position. So we'd need to check the value on every call of `poll`. 
Otherwise we might still have stored an outdated position and delete the wrong waker on drop.

Depends on how much having a runtime of `O(self.waker_queue.len())` for unset (which is only called when a stream is dropped) is really a problem. IMO calling `register` and `unset_waker` is not in the hot code path as these are operations only performed once per lifetime of a stream consumer. But `wake_all` gets called all the time.